src/v/s3/client.h

```diff
@@ -61,6 +61,48 @@ class object_tag_formatter {
     ss::sstring _tags;
 };
 
+enum class error_outcome {
+    none = 0,
+    /// Error condition that could be retried
+    retry,
+    /// The service asked us to retry (SlowDown response)
+    retry_slowdown,
+    /// Error condition that couldn't be retried
+    fail,
+    /// NotFound API error (only suitable for downloads)
+    notfound
```

Comment:
While we're looking at this, I wonder if we should consider differentiating 404s on keys (`notfound`) from other 404s like NoSuchBucket?  Currently we don't have anything smart to do with the latter, but we might in future (e.g. bubbling up a health status if S3 is misconfigured)

Reply:
Ended up adding a `bucket_not_found` error outcome and printing a nasty log message (would be a good candidate for an alert). See https://github.com/redpanda-data/redpanda/pull/7706/commits/e3ef1b1b22504650dc181007ab07b16d5995acc7.